feat: next steps for version management progress (#606)
nodejs/package-maintenance: docs/version-management/proposal-next-steps.md

Proposal: Next steps

As part of achieving the second [goal](./goals.md):

> Install Node.js and a package manager for a local development environment.

And following up on the [proposal to revise the downloads page](./proposal-revise-downloads-page.md), we propose the following next steps:

1. We should revise the Node.js download page to split apart the operating system package managers (Homebrew and Chocolatey) onto their own tab separate from the Node.js version managers (nvm and fnm) and the version managers tab should remain the default. This will further nudge users toward our recommendation of installing Node.js in a version-managed way.

> 2. Also on the download page, we should add instructions for installing Yarn and `pnpm` as package managers to use for a project. These instructions should follow whatever recommendation we receive from those project's maintainers.

Reviewer: What is the criteria behind selecting `yarn` and `pnpm` to be featured in the download page?

Were there discussions on how the decision making is going to work in case new / different package managers wants to be added there?

Author: This is a great point, and one we have not yet discussed. I think the criteria used so far was just "they are anecdotally popular". But this will not stand up over time and we need something more formal.

Do you have suggestions? 

I was thinking we likely need something about ecosystem capability but nailing down language for that is hard. For example, you could say "should be able to install packages from the npm registry and setup a functional program for most existing JS projects in the ecosystem". While I think this is a good one, I could see some folks taking issue with it and having trouble getting consensus.

I was also thinking we could use popularity as a proxy here, and just say "we will only show the top 5 most popular package installers" or something like that, but that also seems like a mine field of issues to defend.

Author: What about something like:

> package managers which node has at any point historically supported, plus others to be determined by a new policy to add and remove recommended package managers in the future.

And then we can add to our todo list to define that policy in a following PR?

I say this because to keep things moving forward, and both of those package managers were supported by `corepack` so seem like good starting points. And we do "pick winners" so it seems reasonable that we could just say "come to the PMWG to propose a new one, if you can get a consensus recommendation we will add it". It is not strict or clear, but it is seemingly reasonable imo.

Reviewer: yeah, adding that part sounds like a good solution for now so that it doesn't prevent progress 👍

3. Corepack's documentation should be moved out of the Node.js API documentation and into its own website, or accessible as Markdown files in the Corepack repo. Corepack is a separate project from Node.js and intermingling its documentation within Node.js's is confusing; we don't do that for `npm` even though we distribute `npm`.

> 
> 4. Once all of the above is complete, we should remove Corepack from the Node.js distribution starting in the next major release. Users who wish to continue using Corepack can do so via the instructions available on the Node.js download page or in Corepack's documentation. This will reduce the maintenance burden on the Node.js project and allow Corepack to evolve independently.

Reviewer 2: Let's remove this part which has not consensus – also "allow Corepack to evolve independently" is not a goal, Corepack should either be superseded or be the standard.
```suggestion
```

Author: > either be superseded or be the standard/recommended way to install package managers.

I think this is the main crux of the problem today right? The package manager authors do not agree on a standard way to install package managers. If they agreed, I would love that outcome but they do not.

Reviewer 2: In any case, "allow Corepack to evolve independently" is not something anyone is asking, nor Corepack developers, nor its users.

Author: I guess I should say, my conclusion arriving at including 4 was that neither of these options are currently viable according to the folks who old decision making power in the respective projects. So I *agree* with you, @aduh95, that would be a much better outcome, but recognizing the reality here have arrived at number 4 being the only current way that both `corepack` and the Node.js project can reasonably both move forward in the way that is best for each project.

Reviewer 2: I'm not talking about what would be a better outcome, I'm saying that:

- "allow Corepack to evolve independently" should not be listed here, it does not have consensus.
- "we should remove Corepack from the Node.js distribution starting in the next major release" should not be listed here, it does not have consensus.

Reviewer 3: > "allow Corepack to evolve independently" should not be listed here, it does not have consensus.

By whom? I assumed the decisions regarding Corepack got deferred to the package maintenance team? Isn't consensus reached even here?

We discussed briefly that Corepac could be moved to the `pkgjs` org

Reviewer 2: If me saying "there is no consensus" is not enough to make it true, I'm not sure what would make it true. The idea of building consensus is not to restrict the list of folks we listen to, but find a compromise that everyone can get behind. "Removing Corepack from the Node.js distribution" without providing alternative is not a compromise that I can get behind, because it's not a compromise, it's a one-sided solution that does not address concerns from folks who use Corepack and would want the project to take ownership of the problems solved by Corepack.

Reviewer 3: Removing it from the Node.js distribution and having it be a standalone package is a solution.

Folks who want to use Corepack can install it with a binary or some other way in the future. Finding a way to bundle all sorts of package managers with their version constraints within Node.js is what feels absurd (due to it's complexity)

So far, I've been seen only drama and chaos related to anything Corepack, and as much as I love using Corepack, it feels like it is misplaced for Node.js a JavaScript runtime to try to make such decisions and encompass all package managers. In my humble opinion, Corepack should not be part of Node.

And to be fair, only you and maybe a handful of more collaborators are interested in insisting that Corepack should be part of Node. Is this a fair ask? I understand that the proposed outcome doesn't make you happy, but aren't you just a _bit_ selfish here, maybe? I don't know how we can address the Corepack issue by "consensus seeking" since you don't seem to be looking for a way of reaching a consensus.

Or so I feel after observing all Corepack-related threads for a while. Apologies for the harsh tone. It's not my intention to be harsh on you, but I genuinely don't know what else can be done here.

Reviewer 2: We can either focus on the points that have consensus, and work on the contentious points as a follow-up – or else what is the plan here? Bully folks who express a different PoV out of the discussion? I reckon finding a compromise is hard here, because despite our best intents we have limited time and are all tired of those Corepack discussions, but in the mean time the status quo should prevail.

(And sure, I'm contributing to Node.js because I'm selfish, and incapable of consensus-seeking, whatever you say)

Reviewer 3: I apologize, Antoine, but my intent here is not to bully you out of the conversation nor to imply that "you contribute to Node" _because_ of being selfish nor that you are incapable of consensus-seeking. Please do not take my words out of context. That's different from my take, and I tried to be as explicit as possible to prevent misunderstandings. I'm not attacking you here; please do not take this personally. I'm just being as honest as possible here.

Truth to be told, although Corepack is part of Node core now, was that a good decision? How many people were involved in such a decision? It's always an excellent self-reflection to re-assess and verify the status quo. Again, please take my assessments with a a whole sack of salt, as I lack a lot of context here, and the least of my intent here is to attack you.

Reviewer 3: Actually, I want to invoke the TSC here to reach a consensus once and for all regarding removing Corepack from being distributed directly within Node.

I've talked with a few collaborators in DM, including Antoine, who explained his reasons for being against what we are proposing here. From what I understood, worse DX was the reason, but what I understand at the moment is that you already need to enable corepack with `corepack enable explicitly`. Moving it outside of Node allows us to install Corepack through `npx corepack` or `npm I -g corepack` or even different ways, which results in a better DX (my personal understanding) + doesn't change existing requirements of having Node (for example if using `npm`)

After reflecting on it for a bit and thinking about consensus-seeking, I do believe that Antoine might be biased as a contributor to Corepack. In light of many of us wanting to find a proper home for Corepack that allows it to shine and be decoupled from Node and its versioning, I humbly request intervention from the TSC, which has already delegated the decision-making for the future of Corepack to the Package Maintenance team.

I respect Antoine's decision to block this change. Still, I find it contradictory that we've been trying to reach a consensus for a vast, unnecessary amount of time, and I'd like to have the TSC input to decide if Antoine's decision should be overridden or not.

Apologies, Antoine for me taking this route, a desperate cry, but also a pretty standard procedure.

cc @nodejs/tsc

Author: I do not believe this thread is "resolved", but I also believe it would be in the best interest of us all if we closed it and re-approached the question of "should point 4 stay or be dealt with after" in a new thread. There are also some other things in here we probably need to discuss, and I will work to extract it into a new set of changes to the proposal doc.

Proposal: Next steps

As part of achieving the second [goal](./goals.md):

> Install Node.js and a package manager for a local development environment.

And following up on the [proposal to revise the downloads page](./proposal-revise-downloads-page.md), we propose the following next steps:

1. We should revise the Node.js download page to split apart the operating system package managers (Homebrew and Chocolatey) onto their own tab separate from the Node.js version managers (nvm and fnm) and the version managers tab should remain the default. This will further nudge users toward our recommendation of installing Node.js in a version-managed way.

2. Also on the download page, we should add instructions for installing package managers which node has at any point historically supported (Yarn and `pnpm), plus any others to be determined by a new policy (yet to be proposed) to add and remove recommended package managers. These instructions should follow whatever recommendation we receive from those project's maintainers.

> 3. Corepack's documentation should be moved out of the Node.js API documentation and into its own website, or accessible as Markdown files in the Corepack repo. Corepack is a separate project from Node.js and intermingling its documentation within Node.js's is confusing; we don't do that for `npm` even though we distribute `npm`.

Reviewer 2: ```suggestion
3. Corepack's documentation should be moved out of the Node.js API documentation and into its own website, or accessible as Markdown files in the Corepack repo. Corepack is a separate project from `node` and intermingling its documentation within `node`'s is confusing; we don't do that for `npm` even though we distribute `npm`.
```

> 4. Once all of the above is complete, we should prepare the eventual removal of Corepack from the Node.js distribution by recommending alternative workflows. Users who wish to continue using Corepack will be recommended to do so via the instructions available on the Node.js download page or in Corepack's documentation.

Reviewer 4: Is the "removal" decision already made?

Reviewer 2: I'm not sure the goal is to "make a decision", but rather to find a plan that can have consensus. If you have objections, or alternative proposals that you think more likely to get consensus, now's the best time to speak!

Reviewer 4: This line makes the decision of removing corepack from node.js. There is no discussion here unless I object. Is that correct?

Reviewer 3: No, I believe you misunderstood. But I think @wesleytodd made updates to this paragraph to better describe that landing this PR != removing Corepack from Node.js

We don't have that power.

Reviewer 2: The premise of this discussion is "Node.js will outlive Corepack", and so the question we're trying to answer is "How can we make it happen in way that maximises everyone's happiness?". If really, you don't see a future for Node.js without Corepack, that's a bummer, but maybe for the sake of the discussion, it's still worth discussing.
IMO Corepack is nothing special, we're more than capable of coming up an alternative that would provide the same DX (or similar enough), also covers Node.js versioning, and ideally would not require getting approval or work from package manager maintainers (which are busy people). If such alternative is able to get consensus, it would make Corepack de facto obsolete; so what's important now is to gather the potential points of contention (surely it's not whether that potential tool is named "Corepack").

Author: @anonrig Am I correct that this line is your primary concern? After the discussion in the main thread, do you have a recommendation to change this line which would resolve your issues? What about something like:

```suggestion
4. Once all of the above is complete, we should start recommending alternative workflows to reduce the churn if/when we decide to remove `corepack` from the Node.js distribution. Users who wish to continue using Corepack will be recommended to do so via the instructions available on the Node.js download page or in Corepack's documentation.
```

Reviewer 5: **suggestion:**

> this PR (on its current state) is not attempting to remove Corepack.

> We don't have that power.

"if/when we decide to remove `corepack`" still reads to me as "eventually we want to remove Corepack" (which makes me interpret the PR as "this is our plan to get rid of Corepack"). So I feel that the text of @wesleytodd's suggestion is still somewhat at odds with the comments I quoted above.

For anyone reading the diff rather than all the comments, I believe this wording change would make clear that this PR is not trying to remove Corepack but is simply recommending improvements for Node package/version management in general.

```suggestion
4. Once all of the above is complete, we should start recommending alternative workflows in case Corepack is removed from the Node.js distribution. Users who wish to continue using Corepack will be recommended to do so via the instructions available on the Node.js download page or in Corepack's documentation.
```

Author: I am perfectly comfortable with this new wording if that achieves agreement to keep the conversation moving forward.

Reviewer 6: remove head from ass or remove npm from world

this debate and PR staying open on eventual removal of corepack which is clearly used by community, is a waste of time from the get go, and a showcase on how out of touch and misprioritized Node.js has gotten

politicized wordsmithing, really?

Reviewer 4: I'm comfortable with @stephenwade's recommendation as well

Author: Perfect. Committed that, are you willing to move your change requested review to an approval?

1. We should revise the Node.js download page to split apart the operating system package managers (Homebrew and Chocolatey) onto their own tab separate from the Node.js version managers (nvm and fnm) and the version managers tab should remain the default. This will further nudge users toward our recommendation of installing Node.js in a version-managed way.

> 2. Also on the download page, we should add instructions for installing package managers which node has at any point historically supported (Yarn and `pnpm), plus any others to be determined by a new policy (yet to be proposed) to add and remove recommended package managers. These instructions should follow whatever recommendation we receive from those project's maintainers.

Reviewer 2: ```suggestion
2. Also on the download page, we should add instructions for installing package managers which node has at any point historically supported (Yarn and `pnpm`), plus any others to be determined by a new policy (yet to be proposed) to add and remove recommended package managers. These instructions should follow whatever recommendation we receive from those projects' maintainers.
```

3. Corepack's documentation should be moved out of the Node.js API documentation and into its own website, or accessible as Markdown files in the Corepack repo. Corepack is a separate project from Node.js and intermingling its documentation within Node.js's is confusing; we don't do that for `npm` even though we distribute `npm`.

4. Once all of the above is complete, we should prepare the eventual removal of Corepack from the Node.js distribution by recommending alternative workflows. Users who wish to continue using Corepack will be recommended to do so via the instructions available on the Node.js download page or in Corepack's documentation.

> 5. Corepack governence and maintenence will be relinquished to the Corepack team who can decide on next steps for the project.

Reviewer 2: I find the phrasing “relinquished to the Corepack team” not very clear, arguably Corepack governance and maintenance are already handled by the Corepack team. I think what we should say is something like:
```suggestion
5. At that point, Corepack would no longer serve a purpose in Node.js project, so its repository will either be archived or transferred.
```

Author: I was trying to make less strong of a stance, but as long as you are alright with this phrasing I am fine with it. Lets maybe leave this open and get a bit of feedback from people on it?

Reviewer 2: I guess we could remove that step altogether, as I said, the Corepack team is already in charge of the governance and maintenance of Corepack, so that step is not adding much.
```suggestion
```
